### PR TITLE
Enable buidler evm to support exact timestamp testing

### PR DESCRIPTION
**Pull request by tranvictor**
**Changes:**

1. Allow evm_mine to accept a timestamp param to mine an empty block at the exact timestamp.
2. Add new evm_setNextBlockTimestamp to register a timestamp for the next block to be mined at.
3. Add the doc for evm_setNextBlockTimestamp.
4. Add relevant tests for those changes.

**Noted behaviors:**
Regarding `evm_setNextBlockTimestamp`, it requires one and only one timestamp param. It set the timestamp as a field named `_nextBlockTimestamp` of the `node`, this field is taken and restored as per snapshot operations, just like `blockTimeOffsetInSeconds`. Once the `_nextBlockTimestamp` is set, the next block will be mined with the timestamp set to its value, EXCEPT if the next block is mined with `evm_mine(timestamp`. Whenever a new block is mined, `_nextBlockTimestamp` is reset to 0 as its null state.

**Usecase:**
For smart contracts that have block timestamp logic, it is necessary to test its edge cases where a transaction is mined at an exact timestamp.

### CODE REVIEW

File: packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts [hunk 1]
```diff
@@ -47,12 +79,37 @@ export class EvmModule {
...
+    }
+    await this._node.increaseTime(increment);
+  }
+
+  private async _mineAction(timestamp: number): Promise<string> {
+    if (timestamp !== 0) {
+      await this._advanceTimeOffsetAccordingToTimestamp(new BN(timestamp));
+    } else {
+      const nextBlockTimestamp = await this._node.getNextBlockTimestamp();
+      if (!nextBlockTimestamp.eq(new BN(0))) {
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts:107]: I don't understand this `if`. Why is this necessary? To ensure that the next block is mined with `nextBlockTimestamp`?

File: packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts [hunk 2]
```diff
@@ -32,6 +41,29 @@ export class EvmModule {
...
+      new BN(latestBlock.header.timestamp)
+    );
+    if (increment.lte(new BN(0))) {
+      throw new InvalidInputError(
+        `Timestamp ${timestamp} is lower than previous block's timestamp` +
+          `${new BN(latestBlock.header.timestamp).toNumber()}`
+      );
+    }
+    await this._node.setNextBlockTimestamp(new BN(timestamp));
+    return timestamp.toString();
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts:64]: Can we return a `numberToRpcQuantity(timestamp)` instead? Most RPC methods encode numbers that way, so I'd rather keep this consistent with them.

File: packages/buidler-core/src/internal/buidler-evm/provider/node.ts [hunk 3]
```diff
@@ -358,6 +360,12 @@ export class BuidlerNode extends EventEmitter {
     await this._saveTransactionAsReceived(tx);
 
     const block = await this._getNextBlockTemplate();
+
+    const timestamp = await this._getNextUsableBlockTimestamp();
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/node.ts:364]: What about moving this logic into `_getNextBlockTemplate`?

This will have two benefits:
- It will decrease code duplication
- It will make the `nextBlockTimestamp` also be used for calls, enabling further control of time there.

Comment (alcuadrado) [reply packages/buidler-core/src/internal/buidler-evm/provider/node.ts:364]: Note that calls shouldn't reset the next block timestamp.

File: packages/buidler-core/src/internal/buidler-evm/provider/node.ts [hunk 4]
```diff
@@ -1044,7 +1069,7 @@ export class BuidlerNode extends EventEmitter {
         header: {
           gasLimit: this._blockGasLimit,
           nonce: "0x42",
-          timestamp: await this._getNextBlockTimestamp()
+          timestamp: await this._getNextUsableBlockTimestamp()
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/node.ts:1072]: Now that I see this here. Shouldn't this be enough?

re: https://github.com/nomiclabs/buidler/pull/508/files#r403510853

File: packages/buidler-core/src/internal/buidler-evm/provider/node.ts [hunk 5]
```diff
@@ -1234,6 +1266,10 @@ export class BuidlerNode extends EventEmitter {
     return latestBlockTimestamp.eq(blockTimestamp);
   }
 
+  private async _setBlockTimestamp(block: Block, timestamp: BN) {
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/node.ts:1269]: Also, this may not be necessary. As it's already set in the block template, right?

File: packages/buidler-core/src/internal/buidler-evm/provider/node.ts [hunk 6]
```diff
@@ -714,6 +737,7 @@ export class BuidlerNode extends EventEmitter {
       latestBlock: await this.getLatestBlock(),
       stateRoot: await this._stateManager.getStateRoot(),
       blockTimeOffsetSeconds: new BN(this._blockTimeOffsetSeconds),
+      nextBlockTimestamp: new BN(this._nextBlockTimestamp),
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/node.ts:740]: You included it in the snapshopts! this is so good 🙌

File: packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts [hunk 7]
```diff
@@ -32,6 +41,29 @@ export class EvmModule {
     throw new MethodNotFoundError(`Method ${method} not found`);
   }
 
+  // evm_setNextBlockTimestamp
+
+  private _setNextBlockTimestampParams(params: any[]): [number] {
+    return validateParams(params, t.number);
```
Comment (alcuadrado) [inline packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts:47]: Oh, also here. I'd use a quantity value.

File: packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts (hunk 1 above)
Comment (tranvictor) [reply packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts:107]: It checks to ensure `nextBlockTimestamp` is set. If it is not (value is 0), it means the next block time Is set and it set the next block timestamp to `nextBlockTimestamp`.

File: packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts (hunk 7 above)
Comment (tranvictor) [reply packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts:47]: This is to keep the consistency with `_increaseTimeParams` API. It would be best if we can support both hex and base 10 params here but I think `number` would be ok for now.

File: packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts (hunk 2 above)
Comment (tranvictor) [reply packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts:64]: This is to be consistent with `_increaseTimeAction` as well.

File: packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts (hunk 7 above)
Comment (tranvictor) [reply packages/buidler-core/src/internal/buidler-evm/provider/modules/evm.ts:47]: It checks to ensure `nextBlockTimestamp` is set. If it is not (value is 0), it means the next block time Is set and it set the next block timestamp to `nextBlockTimestamp`.

File: packages/buidler-core/src/internal/buidler-evm/provider/node.ts (hunk 5 above)
Comment (tranvictor) [reply packages/buidler-core/src/internal/buidler-evm/provider/node.ts:1269]: It is still needed as `mineEmptyBlock(timestamp)` needs to set the block's timestamp (overriding the `nextBlockTimestamp`) if timestamp param is not 0.

File: packages/buidler-core/src/internal/buidler-evm/provider/node.ts (hunk 3 above)
Comment (tranvictor) [reply packages/buidler-core/src/internal/buidler-evm/provider/node.ts]: I updated the code and removed some redundant check-and-set `nextBlockTimestamp`. Now, `_getNextUsableBlockTimestamp` is used by the block template only and will have effects with `runCall` as well.
I still keep the function because it would be too big if we flatten it in block template function. This remains a nice separation of the template logic and the time logic.
